optimum/habana/transformers/models/mpt/modeling_mpt.py

```diff
@@ -21,153 +21,308 @@
+class GaudiMptBlock(MptBlock):
+    def __init__(self, config: MptConfig):
+        super().__init__(config)
+        self.attn = GaudiMptAttention(config)
 
-    residual = hidden_states
+    def allocate_kv_cache(self, batch_size, max_seq_len, inp_seq_len):
+        self.attn.allocate_kv_cache(batch_size, max_seq_len, inp_seq_len)
 
-    # Self attention.
-    attn_outputs, attn_weights, past_key_value = self.attn(
-        layernorm_output,
-        position_bias=position_bias,
-        attention_mask=attention_mask,
-        past_key_value=layer_past,
-        token_idx=token_idx,
-    )
+    def reorder_kv_cache(self, beam_idx: torch.LongTensor):
+        return self.attn.reorder_kv_cache(beam_idx)
 
-    hidden_states = self.resid_attn_dropout(attn_outputs) + residual
+    def update_sincos_cache(self, seq_len):
+        self.attn.update_sincos_cache(seq_len)
 
-    layernorm_output = self.norm_2(hidden_states)
+    def forward(
+        self,
+        hidden_states: torch.Tensor,
+        position_bias: torch.Tensor,
+        attention_mask: torch.Tensor,
+        layer_past: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
+        use_cache: Optional[bool] = False,
+        output_attentions: Optional[bool] = False,
+        token_idx: Optional[torch.Tensor] = None,
+        reuse_cache: Optional[bool] = False,
+        cache_idx: int = None,
+    ):
+        """
+        Copied from MptBlock.forward: https://github.com/huggingface/transformers/blob/v4.32.0/src/transformers/models/mpt/modeling_mpt.py
+        The only differences are:
+        - add new args token_idx
+        - add new args reuse_cache
+        - add new args cache_idx
+        """
+        # hidden_states: [batch_size, seq_length, hidden_size]
+        # Layer norm at the beginning of the transformer layer.
+        layernorm_output = self.norm_1(hidden_states)
+
+        residual = hidden_states
+
+        # Self attention.
+        attn_outputs, attn_weights, past_key_value = self.attn(
+            layernorm_output,
+            position_bias=position_bias,
+            attention_mask=attention_mask,
+            past_key_value=layer_past,
+            use_cache=use_cache,
+            token_idx=token_idx,
+            reuse_cache=reuse_cache,
+            cache_idx=cache_idx,
+        )
 
-    # Get residual
-    residual = hidden_states
+        hidden_states = self.resid_attn_dropout(attn_outputs) + residual
 
-    # MLP.
-    output = self.ffn(layernorm_output, residual)
-    outputs = (output,)
+        layernorm_output = self.norm_2(hidden_states)
 
-    if use_cache:
-        outputs += (past_key_value,)
+        # Get residual
+        residual = hidden_states
 
-    if output_attentions:
-        outputs += (attn_weights,)
+        # MLP.
+        output = self.ffn(layernorm_output, residual)
+        outputs = (output,)
 
-    return outputs  # hidden_states, present, attentions
+        if use_cache:
+            outputs += (past_key_value,)
```

Comment:
It looks like there is some missing processing here. From what I understand, when 'reuse_cache==True', 'past_key_value' stores the shape of the corresponding variable instead of the content itself(line168).

Reply:
Hi @Jing1Ling, Thank you for checking. The section L166 ~ L170 is unnecessary. I removed the section.